pkgs/build-support/rust/rust-utils.nix

```diff
@@ -0,0 +1,374 @@
+                   crateFeatures, libName, build, release, libPath,
+                   crateType, metadata, crateBin, finalBins,
+                   verbose, colors }:
+
+      let depsDir = lib.concatStringsSep " " dependencies;
+          completeDepsDir = lib.concatStringsSep " " completeDeps;
+          completeBuildDepsDir = lib.concatStringsSep " " completeBuildDeps;
+          makeDeps = dependencies:
+            (lib.concatMapStringsSep " " (dep:
+              let extern = lib.strings.replaceStrings ["-"] ["_"] dep.libName; in
+              (if dep.crateType == "lib" then
+                 " --extern ${extern}=${dep.out}/lib/lib${extern}-${dep.metadata}.rlib"
+              else
+                 " --extern ${extern}=${dep.out}/lib/lib${extern}-${dep.metadata}${buildPlatform.extensions.sharedLibrary}")
+            ) dependencies);
+          deps = makeDeps dependencies;
+          buildDeps = makeDeps buildDependencies;
+          optLevel = if release then 3 else 0;
+          rustcOpts = (if release then "-C opt-level=3" else "-C debuginfo=2");
+          rustcMeta = "-C metadata=${metadata} -C extra-filename=-${metadata}";
+          version_ = lib.splitString "-" crateVersion;
+          versionPre = if lib.tail version_ == [] then "" else builtins.elemAt version_ 1;
+          version = lib.splitString "." (lib.head version_);
+      in ''
+      norm=""
+      bold=""
+      green=""
+      boldgreen=""
+      if [[ "${colors}" -eq "always" ]]; then
+        norm="$(printf '\033[0m')" #returns to "normal"
+        bold="$(printf '\033[0;1m')" #set bold
+        green="$(printf '\033[0;32m')" #set green
+        boldgreen="$(printf '\033[0;1;32m')" #set bold, and set green.
+      fi
+
+      echo_build_heading() {
+       start=""
+       end=""
+       if [[ x"${colors}" -eq x"always" ]]; then
+         start="$(printf '\033[0;1;32m')" #set bold, and set green.
+         end="$(printf '\033[0m')" #returns to "normal"
+       fi
+       if (( $# == 1 )); then
+         echo "$start""Building $1""$end"
+       else
+         echo "$start""Building $1 ($2)""$end"
+       fi
+      }
+
+      noisily() {
+        start=""
+        end=""
+        if [[ x"${colors}" -eq x"always" ]]; then
+          start="$(printf '\033[0;1;32m')" #set bold, and set green.
+          end="$(printf '\033[0m')" #returns to "normal"
+        fi
+	${lib.optionalString verbose ''
+            echo -n "$start"Running "$end"
+            echo $@
+	''}
+	$@
+      }
+
+      symlink_dependency() {
+      # $1 is the nix-store path of a dependency
+        i=$1
+	dest=target/deps
+	if [ ! -z $2 ]; then
+           if [ "$2" = "--buildDep" ]; then
+             dest=target/buildDeps
+           fi
+	fi
+        ln -s -f $i/lib/*.rlib $dest #*/
+        ln -s -f $i/lib/*.so $i/lib/*.dylib $dest #*/
+        if [ -e "$i/lib/link" ]; then
+            cat $i/lib/link >> target/link
+            cat $i/lib/link >> target/link.final
```

Comment:
Usually we keep files needed by nix to propagate build information in `nix-support` rather than common directories like `lib`, where only shared objects are stored.

Reply:
The `link` file also causes unnecessary dependencies to be retained in to the final build closure. I added:

```patch
From 93e1765c25114971b0da4efb459ab1eecdd2c4c2 Mon Sep 17 00:00:00 2001
From: Graham Christensen <graham@grahamc.com>
Date: Mon, 4 Dec 2017 19:49:34 -0500
Subject: [PATCH] remove link file

---
 pkgs/build-support/rust/rust-utils.nix | 1 +
 1 file changed, 1 insertion(+)

diff --git a/pkgs/build-support/rust/rust-utils.nix b/pkgs/build-support/rust/rust-utils.nix
index 72bb3b80492..ae8c977a348 100644
--- a/pkgs/build-support/rust/rust-utils.nix
+++ b/pkgs/build-support/rust/rust-utils.nix
@@ -291,6 +291,7 @@ let buildCrate = { crateName, crateVersion, crateAuthors, buildDependencies,
         mkdir -p $out/bin
         cp -P target/bin/* $out/bin # */
       fi
+      rm $out/lib/link || true
     '';
 in
 
-- 
2.15.0
```

and it successfully compiled almost everything (except for ofborg itself.)

As it stands, carnix produces a 700M closure for 8M of binaries:

```
$ nix-store -q --requisites result
/nix/store/dps6gpjd9vmjylqgjhdbw6kyxfbfssn7-glibc-2.26-75
/nix/store/gvcpjv7iv8h3bmwbb7105lkw0frgm7l7-linux-headers-4.4.10
/nix/store/vm4bhyhqalmk4r6hvmam7qh8r6cn8fwc-glibc-2.26-75-bin
/nix/store/0jd5ffyrfkxcggsk4ida0aldfp7d8wpk-glibc-2.26-75-dev
/nix/store/0qk501pbw5dlq9h407923gd486xwlwjb-zlib-1.2.11
/nix/store/1a7vc07czd11iz7zdg10jg5k5p3wvxcf-gawk-4.2.0
/nix/store/2rcaamc809agalljs6yz8dkn1ql8hgnz-attr-2.4.47
/nix/store/kpm8d40rfhf3g4fq68c07ld9wgdjrrdz-acl-2.2.52
/nix/store/1kigwg3iql5rk9qkjnsqb13460g25gvy-coreutils-8.28
/nix/store/l3a1s7ba3fgg7ydh36wcdbl78yaxnysg-openssl-1.0.2m
/nix/store/3zjr2k52ggjdk9mdj7yy0lb47jp7jsc9-gnumake-4.2.1
/nix/store/yfaabj1r18psijpgbl85n3fyxggi2i3d-gcc-6.4.0-lib
/nix/store/43wgf73arrhvqb7cqhnghs59v7abd3ir-patchelf-0.9
/nix/store/5rz7js5wj37xvxrv9gqsqpfmjnl9gamf-gzip-1.8
/nix/store/wagb4mf29ww4lbjrizw1qw81w2087qj8-pcre-8.41
/nix/store/g877mww5r4pv65j7c2dwcym7s01wnbf9-gnugrep-3.1
/nix/store/lphr7qmh21z185zgznvk361bgy8pjvz1-gcc-6.4.0
/nix/store/m37dxy7hjgbsjib1j06bkjbb2w7vjwns-expand-response-params
/nix/store/y6smv5ajhbilpdvf5hlmgbhv7xxfsdpv-binutils-2.28.1
/nix/store/yd2dnycssshrmz4xgximhs3y3d9m1h1f-bash-4.4-p12
/nix/store/8zgagh758g8j35ic6wcgzrpw4ly768ww-gcc-wrapper-6.4.0
/nix/store/d5r2vzgvbx4f9qy5mg76hnk96aynsxsq-ncurses-6.0-20170902-man
/nix/store/v0030l7p0fm0m8h96mj1jy0bg9p8sa0r-ncurses-6.0-20170902
/nix/store/9s16vkbl97a9znwr1yq1sbkddlqivb2k-ncurses-6.0-20170902-dev
/nix/store/gzijxhpkq49npkiq4v9svks7x23g05bs-gnutar-1.29
/nix/store/7cxak3vxiy804kskpxcyhmnp9wcyf79v-bzip2-1.0.6.0.1
/nix/store/hsph80c754i513iid3ypk1vj7qg3fr41-bzip2-1.0.6.0.1-bin
/nix/store/c75y6g7p0kanvy4caapllh7lhii7l4z4-xz-5.2.3
/nix/store/jyhswhm1gajhmqdxchq2d89bid6nmrxi-xz-5.2.3-bin
/nix/store/mb74sc0ipkyz4gycg6rnvx9y5afaaa74-paxctl-0.9
/nix/store/mgav0hzh9y46i1mgq6vypjfq3s41py2x-gnused-4.4
/nix/store/nnprdh4rhgwx1iqv9admr87p7142ds8n-diffutils-3.6
```


```
$ du -hc $(nix-store -q --requisites result)
...
671M	total
```

It'd be really good to figure out a way to only pass on linker flags when strictly required, and maybe a way to shed them on final build products too. In the mean time, I've just deleted the link file in a final cleanup drv.

Reply:
I used the following to strip extra deps:

```nix
let
  stripDeps = pkg: pkgs.runCommand "${pkg.name}-deps-stripped" {}
  ''
    cp -r ${pkg} $out
    chmod -R a+w $out
    rm -rf $out/lib
    find $out/bin -name '*.d' -delete
    chmod -R a-w $out
  '';
in {
  ofborg.rs = stripDeps (pkgs.callPackage ./nix/ofborg-carnix.nix {}).ofborg_0_1_0;
}
```

and it produced:

```
$ nix-store -q --requisites result
/nix/store/dps6gpjd9vmjylqgjhdbw6kyxfbfssn7-glibc-2.26-75
/nix/store/l3a1s7ba3fgg7ydh36wcdbl78yaxnysg-openssl-1.0.2m
/nix/store/g2sg5zl2m7zm2c5fb33ppgvzaiv17a4m-rust_ofborg-0.1.0-deps-strippe

$ du -hc $(nix-store -q --requisites result)
35M	total

```

Reply:
Perhaps this would be a good use case for `dev` outputs? I'm not sure.